alf/agents/simple_agent.py

```diff
@@ -0,0 +1,150 @@
+                 debug_summaries=False,
+                 summarize_grads_and_vars=False,
+                 train_step_counter=None,
+                 name=None):
+        """
+        Args:
+          algorithm (SimpleAlgorithm):
+          iterations_per_batch (int): number of training interations for each
+            training batch
+          gradient_clipping (None|0-D Tensor): Norm length to clip gradients.
+          debug_summaries (bool): If true, subclasses should gather debug
+            summaries.
+          summarize_grads_and_vars (bool): If true, subclasses should
+            additionally collect gradient and variable summaries.
+          train_step_counter (Tensor): An optional counter to increment every
+            time the train op is run.  Defaults to the global_step.
+          name (str): The name of this agent. All variables in this module will
+            fall under that name. Defaults to the class name.
+        """
+        tf.Module.__init__(self, name=name)
+        self._algorithm = algorithm
+        self._is_rnn_policy = bool(algorithm.policy_state_spec)
+        self._iterations_per_batch = iterations_per_batch
+        self._gradient_clipping = gradient_clipping
+        super(SimpleAgent, self).__init__(
+            algorithm.time_step_spec,
+            algorithm.action_spec,
+            policy=algorithm.policy,
+            collect_policy=algorithm.collect_policy,
+            train_sequence_length=None if self._is_rnn_policy else 2,
+            debug_summaries=debug_summaries,
+            summarize_grads_and_vars=summarize_grads_and_vars,
+            train_step_counter=train_step_counter)
+
+    def _split_experience(self, trajectory):
+        next_trajectory = tf.nest.map_structure(lambda x: x[:, 1:], trajectory)
+        trajectory = tf.nest.map_structure(lambda x: x[:, :-1], trajectory)
+
+        # Remove time dim if we are not using a recurrent network.
+        if not self._is_rnn_policy:
+            trajectory, next_trajectory = tf.nest.map_structure(
+                lambda x: tf.squeeze(x, [1]), (trajectory, next_trajectory))
+
+        return trajectory, next_trajectory
+
+    # Use @common.function in graph mode or for speeding up.
+    def _train(self, experience, weights):
+        trajectory, next_trjectory = self._split_experience(experience)
+        for _ in range(self._iterations_per_batch):
+            loss_info = self._train_iter(trajectory, next_trjectory, weights)
+        return loss_info
+
+    def _train_iter(self, trajectory, next_trajectory, weights):
+        batch_size = trajectory.discount.shape[0]
+
+        if self._is_rnn_policy:
+            critics, _ = run_rnn(self._algorithm.pre_critic, next_trajectory,
+                                 self._algorithm.get_initial_state(batch_size))
+        else:
+            critics, _ = self._algorithm.pre_critic_step(next_trajectory)
+        critics = self._algorithm.critic(trajectory, next_trajectory, critics)
+
```

Comment:
I think this split of pre_critic_step and critic is neat and convenient for value-based approaches and most actor-critic approaches, my minor concern comes from the following two types of approaches:
1. GAE computation seems not that straightforward in this split?
2. Do we want to have another base Agent&Algorithm for model-based approaches?